### DISCUSSION TITLE
Player resource props undefined when player leaves game

### DISCUSSION THREAD

**Discussion by ghost** [saul/demofile#264] [Questions] [answered]
**Describe the bug**
Result that becomes undefined if the player leaves the game.

**Code to reproduce**
```js
function getPlayerIndex() {
			playerIndex = demoFile.players.map(p => p.steamId === "BOT" ? p.steamId : new SteamID(p.steamId).getSteamID64()).indexOf(sid.getSteamID64());
			
			playerIndex2 = playerIndex+1;
			let str = playerIndex2.toString().padStart(3, "0")
			player_rank = demoFile.entities.playerResource.props.m_iCompetitiveRanking[str];
			player_wins = demoFile.entities.playerResource.props.m_iCompetitiveWins[str];
			console.log('rank: ' + player_rank + ' wins: ' + player_wins);
}
```

**Output**
```
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 16 wins: 598
rank: 0 wins: 0
rank: 0 wins: 0
rank: 0 wins: 0
```

**Expected behaviour**
Save the previous value if the player quits the game.

**Affected demos**
[Demo](http://replay195.valve.net/730/003409694791636615234_1475086756.dem.bz2) with this [SteamID](https://steamcommunity.com/profiles/76561198314281370)

**Comment by saul** [comment, marked as answer]
Player resource props are cleared when a player disconnects. It's worth noting that the library doesn't massage/aggregate data at all - the data you're reading is coming straight from the demo itself. With that in mind, this behaviour is expected.

If you need to keep track of information before a user disconnects, save it before the user entity is deleted - `demoFile.entities.on('beforeremove', ...)`.

Does this help?

**Comment by saul** [comment]
You will have better performance if you only make _one_ `beforeremove` handler. You need to check which entity is about to be removed by inspecting the properties on `event`. Your code currently has a bug.
